<?xml version="1.0" encoding="utf-8"?>
<feed xmlns="http://www.w3.org/2005/Atom">

  <title><![CDATA[Category: Ios | Thinking...]]></title>
  <link href="http://kaihaodir.github.io/blog/categories/ios/atom.xml" rel="self"/>
  <link href="http://kaihaodir.github.io/"/>
  <updated>2016-03-13T17:58:49+08:00</updated>
  <id>http://kaihaodir.github.io/</id>
  <author>
    <name><![CDATA[Kai]]></name>
    <email><![CDATA[kaihaodir@gmail.com]]></email>
  </author>
  <generator uri="http://octopress.org/">Octopress</generator>

  
  <entry>
    <title type="html"><![CDATA[Arc下NSException有可能会内存泄露]]></title>
    <link href="http://kaihaodir.github.io/blog/2016/03/13/arc-nsexception-cause-memory-leak/"/>
    <updated>2016-03-13T17:00:52+08:00</updated>
    <id>http://kaihaodir.github.io/blog/2016/03/13/arc-nsexception-cause-memory-leak</id>
    <content type="html"><![CDATA[<p>在使用Instrumnents对程序进行内存测试的时候，发现有几处异常处理的地方提示内存泄露。经过分析得知，<strong>因为异常处理，改变代码执行路径，导致编译生成的 release 代码没有执行</strong>。</p>

<p>简单写了些测试代码验证：</p>

<pre><code>- (void)viewDidLoad {

    [super viewDidLoad];

    @try {
        NSMutableArray *stringList = [[NSMutableArray alloc] init];
        for (int i = 0; i &lt; 100; i++) {
            [stringList addObject: @"string instance."];
        }
        [self throwException];
    }
    @catch (NSException *exception) {
        NSLog(@"catch Exception : %@", exception);
    }
    @finally {

    }
}

- (void)throwException
{
    [NSException raise:@"An Exception." format:@"Exception Description."];
}
</code></pre>

<p>反复进出这个页面几次后，在Instruments上果然看到提示内存泄露，而且泄露的地方就指明是 <code>@try{ }</code> 里面创建的对象。</p>

<p><img src="./exception-memory-leak.png" alt="img" /></p>
]]></content>
  </entry>
  
</feed>
